Separate DataFilesDict and MetadataConfigs

lhoestq opened the pull request:
I did a few changes:
- `MetadataConfigs.resolve_data_files_locally` returns a `DataFilesDict` given a config name
- `MetadataConfigs.get_builder_config` returns a `BuilderConfig` given a config name
- `DatasetModule` has optional attributes:
  - dataset_infos: to pass`info=` to `builder_cls()` if the requested config has a dataset_info
  - metadata_configs: to update the builder's hash if the requested config has a metadata_config
  - builder_configs: to configure the builder class
  - default_config_name: to configure the builder class
```python

@dataclass
class DatasetModule:
    module_path: str
    hash: str
    builder_kwargs: dict
    dataset_infos: Optional[DatasetInfosDict] = None
    metadata_configs: Optional[MetadataConfigs] = None
    builder_configs: Optional[List[BuilderConfig]] = None
    default_config_name: Optional[str] = None

```
- other minor improvements
  - removed `get_dataset_info_from_dataset_metadata` since we can just use `DatasetInfosDict.from_metadata`
  - removed `get_dataset_info_from_dataset_infos_json` since we need the full dataset_infos now - I reused the old code

Therefore MetadataConfigs is not modified in place anymore, which makes the logic easier to manipulate IMO

lhoestq committed:
separate DataFilesDict and MetadataConfigs

lhoestq committed:
update tests

lhoestq committed:
fix missing "info"

HuggingFaceDocBuilderDev commented:
_The documentation is not available anymore as the PR was closed or merged._

lhoestq committed:
set default config when has only one dataset_info

lhoestq committed:
typing

lhoestq committed:
fix typing

polinaeterna commented on src/datasets/utils/metadata.py:
```diff
@@ -20,6 +19,10 @@
 
 
+if TYPE_CHECKING:
+    from ..builder import BuilderConfig
```
i'm afraid that might raise circular import error

lhoestq replied:
Fortunately `if TYPE_CHECKING` is here to avoid circular imports ;)

Commit:
fix: pass config name to resolve_data_files_locally

Commit:
fix: tests for local module without script

Commit:
Merge branch 'arbitrary-config-parameters-in-meta-yaml' into separate-metadataconfigs-and-data-files

polinaeterna commented on src/datasets/load.py:
```diff
@@ -1678,8 +1652,14 @@ def load_dataset_builder(
     data_dir = builder_kwargs.pop("data_dir", data_dir)
     data_files = builder_kwargs.pop("data_files", data_files)
-    config_name = builder_kwargs.pop("config_name", name)
+    config_name = builder_kwargs.pop("config_name", name or dataset_module.default_config_name or "default")
```
this causes failing of loading datasets with a single config **without** providing config name, see [this test](https://github.com/huggingface/datasets/pull/5331/files#diff-39f545cd21f1ca75281b3f357c296be83c964764f99dc24888f9e1e3f8fa05c6R1139). Because it passes "default" while should pass None in this case.

basically, "default" should be passed only for non-configured packaged loaders (without metadata)

lhoestq replied:
Which test is not passing ? The link you provided is a diff to many changes

lhoestq committed:
update tests

lhoestq committed:
fix: default config=None when creating a builder

lhoestq committed:
add tests

lhoestq commented:
I switched to passing `None` by default instead of `"default"` for the config name @polinaeterna 

I also added some tests. Let me know what you think :)
